src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,168 @@
+
+    if (CFDSelectHandler::scmInvalidFileDescriptor == mCANSocket) {
+        DEVLOG_ERROR("[CAN Layer] Socket can't be created.\n");
+        return e_InitInvalidId;
+    }
+
+    std::strncpy(mIfr.ifr_name, mCANParam.interfaceName.c_str(), IFNAMSIZ - 1);
+    mIfr.ifr_name[IFNAMSIZ - 1] =
+        '\0'; // Ensure that null termination is added if input is too long.
+    auto err = ioctl(mCANSocket, SIOCGIFINDEX, &mIfr);
+    if (err != 0) {
+        DEVLOG_ERROR("[CAN Layer] Socket can be controlled.\n");
+        close(mCANSocket);
+        return eRetVal;
+    }
+
+    mAddr.can_family = AF_CAN;
+    mAddr.can_ifindex = mIfr.ifr_ifindex;
+
+    if (bind(mCANSocket, (struct sockaddr *)&mAddr, sizeof(mAddr)) < 0) {
+        DEVLOG_ERROR("[CAN Layer] Binding of the socket unsuccessful.\n");
+        close(mCANSocket);
+        return eRetVal;
+    }
+
+    this->getExtEvHandler<CFDSelectHandler>().addComCallback(mCANSocket, this);
+    DEVLOG_DEBUG("[CAN Layer] Callback handler added.\n");
+
+    mCANParam.CANId = mCANParam.CANId;
+
+    eRetVal = e_InitOk;
+    return eRetVal;
+}
+
+CCANComLayer::~CCANComLayer() { closeConnection(); }
+
+void CCANComLayer::closeConnection() {
+    if (mCANSocket != CFDSelectHandler::scmInvalidFileDescriptor) {
+        DEVLOG_DEBUG("[CAN Layer] Closing socket\n");
+        close(mCANSocket);
+        DEVLOG_DEBUG("[CAN Layer] Unregister layer\n");
+        this->getExtEvHandler<CFDSelectHandler>().removeComCallback(mCANSocket);
+    }
+}
+
+EComResponse CCANComLayer::sendData(void *paData, unsigned int) {
+    auto *raw_data = static_cast<CIEC_ARRAY *>(
+        &(static_cast<CIEC_ANY **>(paData)[0]->unwrap()));
+
+    size_t size = raw_data->getToStringBufferSize();
+    DEVLOG_DEBUG("[CAN Layer] Size of the array is: %u.\n", size);
+    std::vector<char> buffer(size);
+
+    int nLength = raw_data->toString(buffer.data(), size);
+    if (nLength <= 0) {
+        std::memset(buffer.data(), 0, size);
+        DEVLOG_ERROR("[CAN Layer] Copying of the raw data failed.");
+    }
+
+    size_t offset = 0;
+    while (offset < size) {
+
+        mFrame.len = static_cast<uint8_t>(
+            (size - offset > CAN_MAX_DLEN) ? CAN_MAX_DLEN : (size - offset));
+
+        // Clear old data and copy new chunk
+        std::memset(mFrame.data, 0, CAN_MAX_DLEN);
+        std::memcpy(mFrame.data, buffer.data() + offset, mFrame.len);
+
+        offset += mFrame.len;
+
+        DEVLOG_DEBUG("[CAN Layer] Sending data ...\n");
+        if (write(mCANSocket, &mFrame, sizeof(struct can_frame)) < 0) {
+            DEVLOG_ERROR("CAN send failed: %s\n", strerror(errno));
+            return e_ProcessDataSendFailed;
+        }
+    }
+
+    return e_ProcessDataOk;
+}
+
+EComResponse CCANComLayer::recvData(const void *paData, unsigned int paSize) {
+    return e_ProcessDataOk;
+}
+
+EComResponse CCANComLayer::processInterrupt() { return e_ProcessDataOk; }
+
+void CCANComLayer::setupCANInternals(CParameterParser &parser) {
```

Comment:
`mCANParam` is only being used during initialization which is where `setupCANInternals` is called. So you could make this method return a `mCANParam` object instead of storing it in the class

Reply:
Makes sense, maybe in the future it will be more complicated and the instance needs the parameters. But for now it's fine.